src/lib/Microsoft.Health.Fhir.Ingest/Service/MeasurementEventNormalizationService.cs

```diff
@@ -88,6 +90,13 @@ private async Task StartConsumer(ISourceBlock<EventData> producer, IAsyncCollect
                             (DateTime.UtcNow - evt.SystemProperties.EnqueuedTimeUtc).TotalMilliseconds);
 
                         var token = _converter.Convert(evt);
+                        var tokenString = JsonConvert.SerializeObject(token);
+                        var bytes = Encoding.Unicode.GetBytes(tokenString).Length;
+
+                        _log.LogMetric(

```

Comment:
This will work for data collection but if we want to use this for a billing meter later we will need to think through placement/dedupping charges.

If the message fails after though should we charge for it?  Similarly if we have a compute failure and we reprocess a set of messages again because a failure on our I don't think it is appropriate to charge the customer again.  

Once again good enough for data collection but we probably need to more if we want to use it in a billing meter in the future.

Reply:
That is a good point. I plan on addressing this, but have not done so yet. I will address as part of this PR.

Reply:
Thoughts on what we would use?  Event Sequence number?

Reply:
I will hold off signing off till the changes make into the PR.  Other changes look good, thank you!  If you end up wanting to do the uniqueness change separately happy to sign off on the content of this PR as is now.  Thank you!

Reply:
I was considering a few solutions for de-duplication that cover the scenarios you mentioned.

Scenarios:
(1) Downstream failures that trigger a retry should not rebill for events when attempting a retry.
(2) App restarted/crashed and we re-read events. We don't want to bill for these events again.

Possible solutions for Scenario (1):
(1) Only logging ingress metric if downstream consumption succeeds. (Pros: seems simple? Cons: slightly different implementations for Functions and Console. Possible CPU hit if we need to iterate over the events again).
(2) Passing the metric to some other service/class with a unique guid or id (Event Sequence number + partition number) and having the service deal with deduplication. Could just be a hashmap that gets flushed every once in a while. (Pros: seems like it also could work. Cons: Uses Memory)
(3) Keeping a dictionary of recently logged metrics for each ingress event batches. If the event batch is already in the dictionary then don't log it again. Periodically purge dictionary values after a certain time period. (Cons: complicated, same memory concerns as above)

For OSS, I am thinking of going with solution (1) to only log the ingress metric if downstream consumption succeeds.

For OSS Functions, it does not appear that the NormalizeDeviceData function currently will retry if there is a downstream failure. It appears to logs an exception and continue on with the next batch. However, it may be best to only log the ingress metric if [dataNormalizationService.ProcessAsync](https://github.com/microsoft/iomt-fhir/blob/7794cbcc463e8d26c3097cd5e2243d770f26fe45/src/func/Microsoft.Health.Fhir.Ingest.Host/IomtConnectorFunctions.cs#L76) succeeds, so was thinking of logging the ingress metric right after the "await". The downside is we may need to iterate of the events to get their size. This will be off my default though.
 
For the OSS console based app, we do retry if there is a downstream failure. The retry occurs in [EventConsumerService ](https://github.com/microsoft/iomt-fhir/blob/7794cbcc463e8d26c3097cd5e2243d770f26fe45/src/lib/Microsoft.Health.Events/EventConsumers/Service/EventConsumerService.cs) and we essentially just call the Normalize processor again. In order to only log the ingress metric once, I am thinking of moving the ingress logging right after [dataNormalizationService.ProcessAsync](https://github.com/microsoft/iomt-fhir/blob/7794cbcc463e8d26c3097cd5e2243d770f26fe45/src/console/Normalize/Processor.cs#L77) so that it only gets logged if the await succeeds. 

Possible solutions for Scenario (2):
(1) If there is a compute failure or restart we need to persist some sort of state of what ingress metrics we have logged so far. If we periodically push some sort of ingress metric datetime in storage that contains a datetime value of that corresponds to the "last" ingress metric we logged for a particular event. We could then modify the code not to log ingress metrics if the event datetime < event datetime in storage. The "checkpoints" would essentially be like our event checkpoints but perhaps we could update these at a more frequent interval.
(2) Like Scenario 1 we could emit a unique id with the metric and have some downstream system consume all the metrics. We could do a query that excludes duplicate ids.

I think Scenario (2) mainly needs to be addressed for PaaS and not OSS at this time, so I am not planning on addressing in this PR.

Thoughts?

Reply:
I went with solution 1 in the latest commit but am open to other ideas too.

Reply:
Sounds good to me.  Thanks for the detailed break down.  For scenario 2 holding off here makes sense.  I personally like option 2 but it depends on what the billing system will support.